### PR TITLE
Lib-app 5: React Navigation multipage app

### PR DESCRIPTION
**Pull request by bhoomika08**

### CODE REVIEW

File: LibraryApp/package.json [hunk 1]
```diff
@@ -12,10 +12,17 @@
   "dependencies": {
     "@react-native-async-storage/async-storage": "^1.13.3",
     "@react-native-community/checkbox": "^0.5.7",
+    "@react-native-community/masked-view": "^0.1.10",
```
Comment (sawangupta92) [inline LibraryApp/package.json:15]: Please check these if we actually need these libraries

File: LibraryApp/src/components/navigators/auth-navigators.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import BooksList from 'components/books-list';
+import ShowBook from 'components/show-book';
+
+const {textCenter} = Global;
+const {fs18, bold, uppercase} = Typography;
+const {lightBlue, purple} = Colors;
+
+const Stack = createStackNavigator();
+
+const AuthNavigators = () => {
```
Comment (sawangupta92) [inline LibraryApp/src/components/navigators/auth-navigators.js]: rename

File: LibraryApp/src/constants/app-defaults.js [hunk 3]
```diff
@@ -1,6 +1,7 @@
 const APP_DEFAULTS = {
-  booksList: 'booksList',
-  libraryForm: 'form',
+  booksList: 'BooksList',
+  libraryForm: 'Form',
+  showBookDetails: "BookDetails"
 };
 
-export const {booksList, libraryForm} = APP_DEFAULTS;
+export const {booksList, libraryForm, showBookDetails} = APP_DEFAULTS;
```
Comment (sawangupta92) [inline LibraryApp/src/constants/app-defaults.js]: rename

File: LibraryApp/src/components/form.js [hunk 4]
```diff
@@ -68,9 +60,37 @@ class Form extends React.PureComponent {
     this.resetAllFields = this.resetAllFields.bind(this);
   }
 
-  displayBooksList() {
-    const {setActiveTab} = this.props;
-    setActiveTab(booksList);
+  componentDidMount() {
```
Comment (sawangupta92) [inline LibraryApp/src/components/form.js]: use .android.js

**Review by sawangupta92 — CHANGES_REQUESTED** [review]

File: LibraryApp/package.json (hunk 1 above)
Comment (bhoomika08) [reply LibraryApp/package.json:15]: Removed @react-navigation/bottom-tabs & react-native-reanimated unused libraries.

File: LibraryApp/src/components/navigators/auth-navigators.js (hunk 2 above)
Comment (bhoomika08) [reply LibraryApp/src/components/navigators/auth-navigators.js]: Changed path to components/navigators.js
Component Name to Navigators

File: LibraryApp/src/constants/app-defaults.js (hunk 3 above)
Comment (bhoomika08) [reply LibraryApp/src/constants/app-defaults.js]: Changed file name to constants/navigators.js

File: LibraryApp/src/components/form.js (hunk 4 above)
Comment (bhoomika08) [reply LibraryApp/src/components/form.js]: Created both .android.js (for Android Devices) and .ios.js (for IOS Devices)

File: LibraryApp/src/components/books-list.js [hunk 5]
```diff
@@ -38,13 +46,45 @@ class BooksList extends React.Component {
...
     };
+    this.handleBackButtonClick = this.handleBackButtonClick.bind(this);
+    this.setExitApp = this.setExitApp.bind(this);
     this.onRefresh = this.onRefresh.bind(this);
     this.selectBook = this.selectBook.bind(this);
     this.search = this.search.bind(this);
     this.addNewBook = this.addNewBook.bind(this);
   }
 
+  componentDidMount() {
```
Comment (sawangupta92) [inline LibraryApp/src/components/books-list.js]: Is it possible to take out all logic regarding back button feature to another component? books-list component should only be responsible for listing books and nothing else

File: LibraryApp/src/components/books-list.js [hunk 6]
```diff
@@ -38,13 +46,45 @@ class BooksList extends React.Component {
...
+
+  componentWillUnmount() {
+    removeHardwareBackAction(this.handleBackButtonClick);
+  }
+
+  handleBackButtonClick() {
+    const {exitApp} = this.state;
+    setTimeout(() => {
+      this.setExitApp(0);
+    }, 2000); // 2 seconds to tap second-time
```
Comment (sawangupta92) [inline LibraryApp/src/components/books-list.js]: 2 seconds should be a constant

File: LibraryApp/src/components/books-list.js (hunk 5 above)
Comment (sawangupta92) [reply LibraryApp/src/components/books-list.js]: Also make sure that you use functional component with useState hook

File: LibraryApp/src/components/books-list.js [hunk 7]
```diff
@@ -54,9 +94,13 @@ class BooksList extends React.Component {
   }
 
   selectBook(book) {
-    const {setSelectedBook} = this.props;
+    const {setSelectedBook, navigation} = this.props;
     setSelectedBook(book);
-    this.navigateToForm();
+    navigation.navigate(showBookDetails, {
+      params: {
```
Comment (sawangupta92) [inline LibraryApp/src/components/books-list.js]: do we need params?

File: LibraryApp/src/components/form.js [hunk 8]
```diff
@@ -12,36 +13,29 @@ import {
...
   constructor(props) {
     super(props);
-    const {
-      id = '',
-      name = '',
-      author = '',
-      publisher = '',
-      price = '',
-    } = props.selectedBook;
+    const {id = '', name = '', author = '', publisher = '', price = ''} =
```
Comment (sawangupta92) [inline LibraryApp/src/components/form.js]: why are we changing this formatting?

File: LibraryApp/src/components/form.js [hunk 9]
```diff
@@ -57,8 +51,8 @@ class Form extends React.PureComponent {
       errors: {},
       isLoading: false,
     };
-
-    this.displayBooksList = this.displayBooksList.bind(this);
+    this.handleBackButtonClick = this.handleBackButtonClick.bind(this);
```
Comment (sawangupta92) [inline LibraryApp/src/components/form.js]: again, let's try to move all back button related logic into a single file & then we add switch case to handle the rediction

**Review by sawangupta92 — CHANGES_REQUESTED** [review]

File: LibraryApp/src/components/books-list.js (hunk 6 above)
Comment (bhoomika08) [reply LibraryApp/src/components/books-list.js]: Done

File: LibraryApp/src/components/books-list.js (hunk 5 above)
Comment (bhoomika08) [reply LibraryApp/src/components/books-list.js]: Done

File: LibraryApp/src/components/books-list.js (hunk 7 above)
Comment (bhoomika08) [reply LibraryApp/src/components/books-list.js]: Removed. Working on removing params as well.

File: LibraryApp/src/components/form.js (hunk 8 above)
Comment (bhoomika08) [reply LibraryApp/src/components/form.js]: Just destructuring the selectedBook details to use it in state.

File: LibraryApp/src/components/form.js (hunk 9 above)
Comment (bhoomika08) [reply LibraryApp/src/components/form.js]: Done

File: LibraryApp/src/components/form.js [hunk 10]
```diff
@@ -12,36 +13,26 @@ import {
...
-import {booksList} from 'constants/app-defaults';
-import {setActiveTab, updateBooks} from 'store/actions/library';
+import {booksList} from 'constants/navigators';
+import {updateBooks} from 'store/actions/library';
 import {omit, alert} from 'helpers/application';
 import {GetBookDetails} from 'helpers/library';
 import {isEmail, isFieldEmpty, isUrl} from 'helpers/validation';
 import {InputField, Checkbox, Dropdown} from 'components/shared/form-controls';
-import {previous} from 'constants/icons';
+import HandleHardwareBack from 'components/shared/handle-hardware-back';
```
Comment (sawangupta92) [inline LibraryApp/src/components/form.js]: I think in this case we should either HOC or a custom hook. Pls read this guide - https://reactjs.org/docs/hooks-custom.html

Comment (sawangupta92) [reply LibraryApp/src/components/form.js]: let me know in case you need any discussion here

File: LibraryApp/src/components/navigators.js [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+    const route = state.routes[state.index];
+    if (route.state) {
+      return isCurrentScreenInitialOne(route.state);
+    }
+    return state.index === 0;
+  };
+
+  return (
+    <>
+      {isInitialScreen && <HandleHardwareBack />}
```
Comment (sawangupta92) [inline LibraryApp/src/components/navigators.js]: looks complicated. Is it possible to move this into the component itself. As we are going to have a single initialScreen? Pls let me know your thoughts behind this

**Review by sawangupta92 — CHANGES_REQUESTED** [review]

File: LibraryApp/src/components/form.js (hunk 10 above)
Comment (bhoomika08) [reply LibraryApp/src/components/form.js]: Used custom hook for handling hardware back.

File: LibraryApp/src/components/navigators.js (hunk 11 above)
Comment (bhoomika08) [reply LibraryApp/src/components/navigators.js]: Moved this to book-lists component.

**Pull request merged**